### PR TITLE
fix(sql): fix ArrayIndexOutOfBoundsException in select distinct with duplicate columns

### LINKED ISSUE #3833: ArrayIndexOutOfBoundsException: Index -1 out of bounds for length 16
### Describe the bug

```
C i.q.c.p.PGWireServer internal error [ex=
java.lang.ArrayIndexOutOfBoundsException: Index -1 out of bounds for length 16
	at io.questdb.std.ObjList.getQuick(ObjList.java:182)
	at io.questdb.cairo.AbstractRecordMetadata.getColumnMetadata(AbstractRecordMetadata.java:63)
	at io.questdb.cairo.AbstractRecordMetadata.getColumnType(AbstractRecordMetadata.java:73)
	at io.questdb.griffin.SqlCodeGenerator.generateSelectChoose(SqlCodeGenerator.java:3084)
	at io.questdb.griffin.SqlCodeGenerator.generateSelect(SqlCodeGenerator.java:2750)
	at io.questdb.griffin.SqlCodeGenerator.generateQuery0(SqlCodeGenerator.java:2358)
	at io.questdb.griffin.SqlCodeGenerator.generateQuery(SqlCodeGenerator.java:2345)
	at io.questdb.griffin.SqlCodeGenerator.generate(SqlCodeGenerator.java:201)
	at io.questdb.griffin.SqlCompilerImpl.generateFactory(SqlCompilerImpl.java:2819)
	at io.questdb.griffin.SqlCompilerImpl.generate(SqlCompilerImpl.java:2815)
	at io.questdb.griffin.SqlCompilerImpl.compileUsingModel(SqlCompilerImpl.java:1437)
	at io.questdb.griffin.SqlCompilerImpl.compileInner(SqlCompilerImpl.java:1404)
	at io.questdb.griffin.SqlCompilerImpl.compileBatch(SqlCompilerImpl.java:287)
	at io.questdb.cairo.pool.SqlCompilerPool$C.compileBatch(SqlCompilerPool.java:103)
	at io.questdb.cutlass.pgwire.PGConnectionContext.processQuery(PGConnectionContext.java:2483)
	at io.questdb.cutlass.pgwire.PGConnectionContext.parse(PGConnectionContext.java:1902)
	at io.questdb.cutlass.pgwire.PGConnectionContext.handleClientOperation(PGConnectionContext.java:515)
	at io.questdb.cutlass.pgwire.PGWireServer$1.lambda$$0(PGWireServer.java:86)
	at io.questdb.network.AbstractIODispatcher.processIOQueue(AbstractIODispatcher.java:199)
	at io.questdb.cutlass.pgwire.PGWireServer$1.run(PGWireServer.java:124)
	at io.questdb.mp.Worker.run(Worker.java:118)
]

```

### To reproduce

Run: 
`( SELECT DISTINCT count(*), count(*) FROM telemetry as T1 CROSS JOIN telemetry as T2 )`

### Expected Behavior

_No response_

### Environment

```markdown
- **QuestDB version**: 7.3.3
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by puzpuzpuz**
Fixes #3833

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 12 / 12 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|3|3|100.00%|

File: core/src/main/java/io/questdb/griffin/model/QueryModel.java [hunk 1]
```diff
@@ -252,11 +252,6 @@ public void addField(QueryColumn column) {
         columnAliasIndexes.put(alias, bottomUpColumnNames.size() - 1);
     }
 
-    public void addFieldForNullAlias(QueryColumn column) {
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/model/QueryModel.java:255]: Unused in OSS and ent.

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java [hunk 2]
```diff
@@ -2908,10 +2909,10 @@ private void propagateTopDownColumns0(QueryModel model, boolean topLevel, @Nulla
         final QueryModel nested = skipNoneTypeModels(model.getNestedModel());
         model.setNestedModel(nested);
         final boolean nestedIsFlex = modelIsFlex(nested);
-        final boolean nestedAllowsColumnChange = nested != null && nested.allowsColumnsChange();
+        final boolean nestedAllowsColumnChange = nested != null && nested.allowsColumnsChange()
+                && model.allowsNestedColumnsChange();
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/SqlOptimiser.java:2913]: The fix is here.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 14 / 14 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|4|4|100.00%|

File: core/src/main/java/io/questdb/griffin/model/QueryModel.java [hunk 3]
```diff
@@ -302,19 +297,34 @@ public void addUpdateTableColumnMetadata(int columnType, String columnName) {
...
             }
-
             union = union.getUnionModel();
         }
+        return true;
+    }
 
+    /**
+     * Determines whether this model allows pushing columns to nested models.
+     * If this is a SELECT DISTINCT then we should the parent model will contain the necessary columns.
```
Comment (bziobrowski) [inline core/src/main/java/io/questdb/griffin/model/QueryModel.java:318]: typo ?

Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/model/QueryModel.java]: A typo indeed. Fixed in 8d841ae0c1328e2b18583fff5b03a79d91e33fe5

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 14 / 14 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|4|4|100.00%|

**Review by bziobrowski — APPROVED** [review]
:+1: 

I guess in the future we'll need to apply duplicate expression optimization to distinct model.

File: core/src/test/java/io/questdb/test/griffin/DistinctTest.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+package io.questdb.test.griffin;
+
+import io.questdb.test.AbstractCairoTest;
+import org.junit.Test;
+
+/**
+ * These tests cover distinct variations.
+ */
+public class DistinctTest extends AbstractCairoTest {
+
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/test/java/io/questdb/test/griffin/DistinctTest.java:34]: Done in 02b5d20a1bd66dfbb95442557092fb800d85d29e

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 14 / 14 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|4|4|100.00%|

**Review by bziobrowski — APPROVED** [review]

**Comment by puzpuzpuz** [comment]
@bziobrowski thanks for the review!

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 14 / 14 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|4|4|100.00%|

**Pull request merged**